torch/nested/_internal/ops.py

```diff
@@ -884,16 +884,91 @@ def split_with_sizes_default(func, *args, **kwargs):
     )
     inp = new_kwargs.pop("input")
 
-    dim = _wrap_jagged_dim(inp.dim(), new_kwargs["dim"], inp._ragged_idx, "narrow")
+    dim, operating_on_batch = _wrap_jagged_dim(
+        inp.dim(), new_kwargs["dim"], inp._ragged_idx, "narrow", allow_batch_dim=True
+    )
+    if operating_on_batch:
+        # batch dim narrowing requires custom logic involving offsets
+        out_kwargs = extract_kwargs(inp)
+        start_val, length_val = new_kwargs["start"], new_kwargs["length"]
+        end_val = start_val + length_val
+        batch = inp._offsets.shape[0] - 1
+        if end_val > batch:
+            raise RuntimeError(
+                f"narrow(): start ({start_val}) + length ({length_val}) "
+                f"exceeds dimension size ({batch})"
+            )
+
+        # clamp start, end values
+        if start_val < 0:
+            start_val += inp._values.size(dim)
+        if end_val < 0:
+            end_val += inp._values.size(dim)
+        start_val = max(min(start_val, inp._values.size(dim)), 0)
+        end_val = max(min(end_val, inp._values.size(dim)), 0)
+        length_val = max(min(length_val, end_val - start_val), 0)
+
+        # shortcut if no actual narrowing is happening; this helps us ensure
+        # that length < batch size if we don't take this path
+        if length_val == inp.size(0):
+            return inp.detach()
+
+        # +1 to include last offset. Also normalize offsets to start at 0.
+        out_kwargs["offsets"] = (
+            inp._offsets[start_val : start_val + length_val + 1]
+            - inp._offsets[start_val]
+        )
+        # metadata cache may no longer be accurate since offsets have changed
+        if "_metadata_cache" in out_kwargs:
+            del out_kwargs["_metadata_cache"]
+
+        if inp._lengths is not None:
+            out_kwargs["lengths"] = inp._lengths[start_val : start_val + length_val]
+
+        # unbacked SymInt for new storage offset
+        new_storage_offset = (
+            inp._values.storage_offset()
+            + (inp._offsets[start_val] * inp._values.stride(dim))
+        ).item()
+        torch._check_is_size(new_storage_offset)
+
+        # compute symbolic start involving unbacked SymInt
+        start = (
+            new_storage_offset - inp._values.storage_offset()
+        ) // inp._values.stride(dim)
+        torch._check_is_size(start)
+        torch._check(start <= inp._values.size(dim))
+
+        # unbacked SymInt for length
+        length = (inp._offsets[start_val + length_val] - inp._offsets[start_val]).item()
+        torch._check_is_size(length)
+        # we can say this because we short-circuit earlier if length == inp._values.size(dim)
+        torch._check(length < inp._values.size(dim))
+        torch._check(start + length <= inp._values.size(dim))
+
+        # compute new sizes / strides from symbolic values
+        new_sizes = list(inp._values.size())
+        new_sizes[dim] = length
```

Comment:
I guess this + using as_strided instead of narrow is to avoid pending unbacked symbols? Maybe worth adding a small comment :)

Reply:
yep exactly, added a comment describing the situation. Essentially, we need all unbacked symints allocated to be directly accessible from the returned tensor's metadata (sizes, strides, storage offset) and not part of non-trivial expressions (e.g. as you'd get for the storage offset from calling narrow() with unbacked SymInt inputs)